orchagent/aclorch.cpp

```diff
@@ -2183,6 +2262,105 @@ void AclRuleMirror::onUpdate(SubjectType type, void *cntx)
+    AclRule(aclOrch, rule, table, createCounter),
+    table_id(table),
+    m_metaDataMgr(m_metaDataMgr)
+{
+}
+
+uint32_t AclRuleUnderlaySetDscp::getDscpValue() const
+{
+    return cachedDscpValue;
+}
+
+uint32_t AclRuleUnderlaySetDscp::getMetadata() const
+{
+    return cachedMetadata;
+}
+
+bool AclRuleUnderlaySetDscp::validateAddAction(string attr_name, string _attr_value)
+{
+    SWSS_LOG_ENTER();
+
+    string attr_value = to_upper(_attr_value);
+
+    sai_object_id_t table_oid = m_pAclOrch->getTableById(table_id);
+    auto aclTable = m_pAclOrch->getTableByOid(table_oid);
+    string type = aclTable->type.getName();
+    string key = table_id + ":" + m_id;
+    // we handle the allocation of metadata for here. based on SET_DSCP action, we check if a metadata is already allocated then we reuse it
+    // otherwise we allocate a new metadata. This metadata is then set an the action for the Rule of this table. We also cache the SET_DSCP
+    // value and the allocated metadata in a the rule structure itself so that when we go to addRule we can use these to add the
+    // egr_set_dscp rule
+    if (attr_name == ACTION_DSCP && (type == TABLE_TYPE_MARK_META || type == TABLE_TYPE_MARK_META_V6))
+    {
+        if (!m_pAclOrch->isUsingEgrSetDscp(table_id))
+        {
+
+            SWSS_LOG_ERROR("Unexpected Error. Table %s not asssociated with EGR_SET_DSCP table", table_id.c_str());
+            return false;
+        }
+        if (m_pAclOrch->hasMetaDataRefCount(key) > 0)
+        {
+            SWSS_LOG_ERROR("Metadata already allocated for Rule %s in table %s. Remove and Re-add the rule.", m_id.c_str(), table_id.c_str());
+            return false;
+        }
+        u_int8_t actionDscpValue = uint8_t(std::stoi(attr_value));
+        cachedDscpValue = actionDscpValue;
+        auto metadata = m_metaDataMgr->getFreeMetaData(actionDscpValue);
+
+        if (!m_metaDataMgr->isValidMetaData(metadata))
+        {
+            SWSS_LOG_ERROR("Failed to get free metadata for DSCP value %d", actionDscpValue);
+            return false;
+        }
+        SWSS_LOG_ERROR("Allocated metadata %d for DSCP value %d", metadata, actionDscpValue);
+        cachedMetadata = metadata;
+        attr_name = ACTION_META_DATA;
+        attr_value = std::to_string(metadata);
+        m_pAclOrch->addMetaDataRef(key, metadata);
+    }
+
+
+    sai_acl_action_data_t actionData;
+    actionData.parameter.u32 = 0;
+
+    SWSS_LOG_INFO("attr_name: %s, attr_value: %s int val %d", attr_name.c_str(), attr_value.c_str(), to_uint<uint32_t>(attr_value));
+    // we only handle DSCP and META_DATA actions for now.
+    if (attr_name == ACTION_DSCP || attr_name == ACTION_META_DATA)
+    {
+        actionData.parameter.u32 = to_uint<uint32_t>(attr_value);
+        if (attr_name == ACTION_META_DATA && (actionData.parameter.u32 < m_pAclOrch->getAclMetaDataMin() || actionData.parameter.u32 > m_pAclOrch->getAclMetaDataMax()))
+        {
+            return false;
```

Comment:
For any failures, should we decrement the Metadata Ref count??

Reply:
I do the deref on removal.failure results in failure status, when the rule is removed I deref the metadata as well. This check serves as validator and safety net. We hit this when we have exhausted the metadata values or when we have an invalid metadata value. so we cannot proceed.